### PR TITLE
[AIRFLOW-1872] Set context for all handlers including parents

### PR DESCRIPTION
**Pull request by bolkedebruin**
Make sure you have checked _all_ steps below.

### JIRA
- [X] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-1872

### Description
- [X] Here are some details about my PR, including screenshots of any UI changes:

Previously setting the context was not propagated to the parent
loggers.

### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

Will add

### Commits
- [X] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

- [X] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

cc @jgao54 @ashb please take a look

### CODE REVIEW

**Comment by jgao54** [comment]
@bolkedebruin it is not immediately obvious to me what the root issue was, your code look fine to me, but I am uncertain how it fixed the bug (since set_context has always been attempted for 'ariflow.task.raw', it sounded like it's the handlers that were missing from the logger which prevented the logs from being recorded)

**Comment by bolkedebruin** [comment]
@jgao54 I understand. The issue is that the logger being returned seems to be shallow when not explicitly defined. 

```
>>> import airflow.settings
>>> import logging
>>> log = logging.getLogger("airflow.task.raw")
>>> log.handlers
[]
>>> log.parent
<logging.Logger object at 0x1080a5590>
>>> log.parent.handlers
[<airflow.utils.log.file_task_handler.FileTaskHandler object at 0x1090fcc90>]
```

**Comment by jgao54** [comment]
That makes sense.  So if I understood correctly, is the goal to propagate handlers from 'airflow.task.raw' to 'airflow.task'?

This means we'd have to explicitly define propagate=true for 'airflow.task.raw' in the logging config right?

**Comment by criccomini** [comment]
+1

**Comment by bolkedebruin** [comment]
@jgao54 No it is not the goal to propagate the handlers, but making sure to context gets set across all handlers across the tree of loggers (maybe you meant that?). So in case you define airflow.task.raw explicitly and would like to make use of the context you need to define a handler that understand this. In case you allow propagation airflow.task will make use of the context setting _as well_

**Comment by jgao54** [comment]
I'd prefer eventually removing `airflow.task.raw` completely as in #2793 , but this PR cleans up the set context as well so lgtm. Will wait for @ashb to verify the fix.

**Comment by ashb** [comment]
I'll run some tests shortly code diff (from viewing on my phone) looks sensible.

**Comment by ashb** [comment]
Looks like the tests failed for a random timeout related reason.

This fix works for me.

**Comment by jgao54** [comment]
Turns out that logger.propagate by default is set to True, makes sense to me now that this fix works without having to explicitly define 'airflow.task.raw' in the config.

**Comment by bolkedebruin** [comment]
I worked it out a little bit further. I have made Operators use the log of "airflow.task.operators" so it will bubble up to "airflow.task". This means that this:

```
[2017-12-01 11:24:56,874] {test.py:13} ERROR - Hello
[2017-12-01 11:24:56,874] {logging_mixin.py:94} INFO - [2017-12-01 11:24:56,874] {python_operator.py:91} INFO - Done. Returned value was: None
```

becomes:

```
[2017-12-01 12:13:29,086] {test.py:13} ERROR - Hello
[2017-12-01 12:13:29,086] {python_operator.py:91} INFO - Done. Returned value was: None
```

So no double log lines anymore.

**Comment by bolkedebruin** [comment]
@ashb PTAL (don't know yet if tests will pass)

Note: i might need to put the "RedirectStdHandler" back in, to accomodate for users that do

```
import logging

log = logging.getLogger()

```

In their DAGs.

**Comment by ashb** [comment]
Will test it out.

**Comment by bolkedebruin** [comment]
cc @Fokko

File: airflow/utils/log/logging_mixin.py [hunk 1]
```diff
@@ -107,6 +110,34 @@ def isatty(self):
         return False
 
 
+class RedirectStdHandler(StreamHandler):
+    """
+    Handler that allows stdout/stderr to be redirected after initalization
+    """
+    def __init__(self, stream):
+        if isinstance(stream, file):
```
Comment (ashb) [inline airflow/utils/log/logging_mixin.py]: This doesn't work in Python3 btw. Looking up what the alternative is.

Comment (ashb) [reply airflow/utils/log/logging_mixin.py]: https://stackoverflow.com/a/36321030/439189:

```python
import io
try:
    file_types = (file, io.IOBase)
except NameError:
    file_types = (io.IOBase,)

with open("README.md", "r") as fin:
    print(isinstance(fin, file_types))
```

**Comment by ashb** [comment]
I think applying this patch does something funny to the scheduler -- its sitting at 100% CPU and two of the processing children seem to never exit:

```
================================================================================
DAG File Processing Stats

File Path                                            PID  Runtime    Last Runtime    Last Run
-------------------------------------------------  -----  ---------  --------------  ----------
/usr/local/airflow/dags/brand_index.py
/usr/local/airflow/dags/semasio.py                    16  30.24s
/usr/local/airflow/dags/tapad_optout_pooler.py
/usr/local/airflow/dags/tapad_processor.py            15  30.27s
/usr/local/airflow/dags/example/csv_to_parquet.py
================================================================================
```

Under rc4 the runtime is usually ~1s, and Last Run column gets values

**Comment by bolkedebruin** [comment]
@ashb are you able to track where time is being spent?

**Comment by ashb** [comment]
I haven't managed to track down the time yet. Will see if I can spare any time this afternoon

**Comment by bolkedebruin** [comment]
I will fix the other issues separately and target them for 1.9.1. It is taking more time than expected

**Comment by bolkedebruin** [comment]
Will merge (tests work locally, core dump in this one that is unrelated)

**Pull request closed without merging**